posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

```diff
@@ -0,0 +1,414 @@
+
+Because we have a JSON array, we can use `jsonb` to convert the JSON to an `ArrayList` of the keys that are requested from the properties `HashMap`.
+```
+/* Take the Json Array data and use that to pull out the request properties */
+ArrayList<String> tProps = jsonb.fromJson(jsonString, ArrayList.class);
+for(String key: tProps){
+    qProps.put(key, props.get(key));
+}
+```
+We use the other hashmap created at the start to store the properties we queried for.
+
+Now that we have built our map of queried properties, it can be returned to the user in the same way we returned the full list of properties.
+```
+return CloudEventBuilder.v1()
+    .withData(jsonb.toJson(qProps).getBytes())
+    .withDataContentType("application/json")
+    .withId("properties")
+    .withType("java.properties")
+    .withSource(URI.create("http://system.poperties"))
+    .build();
+```
+
+Next, we're ready to test the new method.
+
+To invoke the method, we make a POST request against `/dev/system/properties` with the HTTP request being a CloudEvent. You can do this locally, or against a rebuilt Docker image.
+
+To invoke this method, use the following curl command:
+```
+curl -X POST http://${url}/dev/system/properties \
+-H "Ce-Specversion: 1.0" \
+-H "Ce-Type: properties" \
+-H "Ce-Source: io.cloudevents.examples/properties" \
+-H "Ce-Id: 536808d3-88be-4077-9d7a-a3f162705f78" \
+-H "Content-Type: application/json" \
+-H "Ce-Subject: resources" \
+-d "[\"java.vendor.url\",\"awt.toolkit\"]"
+```
+
+In the same way that we returned a CloudEvent, when we make the request, we need to provide the required set of headers so that the application can correctly convert the request into a CloudEvent.
+
+=== Updating your IBM Cloud Code Engine application
+The application can now return and receive CloudEvents.  We can update our application in ICCE.
+
+To update your application, complete the following steps:
+
+1. Rebuild your Liberty application with the CloudEvent changes.
+2. Rebuild your docker container and publish to it ICCR, either by updating the image tag or leaving it as is if you are using the image hash
+3. https://cloud.ibm.com/docs/codeengine?topic=codeengine-update-app#update-app-crimage-console[Update your ICCE application] to use the new version.
+
+After you update your application, you can validate your changes in ICCE by invoking the same curl commands that we used locally, but replacing the protocol and hostname.
+```
+curl -X POST https://${ICCE_Application_URL}/dev/system/properties \
+-H "Ce-Specversion: 1.0" \
+-H "Ce-Type: properties" \
+-H "Ce-Source: io.cloudevents.examples/properties" \
+-H "Ce-Id: 536808d3-88be-4077-9d7a-a3f162705f78" \
+-H "Content-Type: application/json" \
+-H "Ce-Subject: resources" \
+-d "[\"java.vendor.url\",\"awt.toolkit\"]"
+```
+
+You now have a MicroProfile application on Open Liberty running on IBM Cloud Code Engine that handles CloudEvents.
```

Comment:
this is better- but it should mention/reiterate  "serverless" in the result. Anything we can say/show to sum up the value  added by going through this process. Right now it is focused on the tools, how can we emphasize the outcome or benefits?

Reply:
I agree, was working on an alternative

its a bit of a mouthful, but this covers everything that they has been done throughout the post:
`You now have a MicroProfile application on Open Liberty running as a serverless function in IBM Cloud Code Engine with CloudEvents!`

Reply:
Thanks that looks good. With this outcome- can/does the ICCE allocate resources to the app on demand scale it to zero when not in use?

Reply:
The App Startup and  Scale to Zero are covered at the end of the initial App deployment section, before they move onto CloudEvents. given we are also trying to promote ICCE, I think this would be good to move to the end so that it is not lost in the rest of the post.

Reply:
great idea